src/coprocessor/dag/batch/executors/limit.rs

```diff
@@ -0,0 +1,304 @@
+                assert!(res.is_drained.as_ref().unwrap());
+                assert_eq!(res.data.rows_len(), limit);
+            } else {
+                assert!(!res.is_drained.as_ref().unwrap());
+                assert_eq!(res.data.rows_len(), get_rows);
+            }
+        }
+    }
+
+    #[test]
+    fn test_execution_summary() {
+        let src = MockExecutor::new();
+        let mut executor =
+            BatchLimitExecutor::new(src, 4, ExecSummaryCollectorEnabled::new(1)).unwrap();
+        executor.next_batch(1);
+        executor.next_batch(2);
+        let mut s = BatchExecuteStatistics::new(2, 1);
+        // Collected statistics remain unchanged until `next_batch` generated delta statistics.
+        for _ in 0..2 {
+            executor.collect_statistics(&mut s);
+            let exec_summary = s.summary_per_executor[1].as_ref().unwrap();
+            assert_eq!(3, exec_summary.num_produced_rows);
+            assert_eq!(2, exec_summary.num_iterations);
+        }
+
+        // we get 1 row since the limit is 4
+        executor.next_batch(10);
+        //s.clear();
+        executor.collect_statistics(&mut s);
+        let exec_summary = s.summary_per_executor[1].as_ref().unwrap();
+        assert_eq!(4, exec_summary.num_produced_rows);
+        assert_eq!(3, exec_summary.num_iterations);
+    }
+
+    #[test]
+    fn test_invalid_limit() {
+        let src = MockExecutor::new();
+        assert!(BatchLimitExecutor::new(src, 0, ExecSummaryCollectorDisabled).is_err());
+    }
+
+    /// MockErrExecutor is based on MockExecutor, the only difference is
+    /// that when call the function next_batch, it always returns is_drained error.
+    struct MockErrExecutor(MockExecutor);
+    impl MockErrExecutor {
+        fn new() -> Self {
+            MockErrExecutor(MockExecutor::new())
+        }
+    }
+    impl BatchExecutor for MockErrExecutor {
+        #[inline]
+        fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+            let mut result = self.0.next_batch(expect_rows);
+            result.is_drained = Err(box_err!("next batch mock error"));
+            result
+        }
+
+        #[inline]
+        fn collect_statistics(&mut self, _: &mut BatchExecuteStatistics) {}
+
+        #[inline]
+        fn schema(&self) -> &[FieldType] {
+            self.0.schema()
+        }
+    }
+
+    #[test]
+    fn test_src_next_batch_err() {
+        let data = vec![
+            //limit, expect_rows(real get rows)
+            (4, 3), // error happens before limit rows
+            (3, 4), // error happens after limit rows
+            (3, 3), // error happens when get the limit + 1 row.
+        ];
+        for (limit, get_rows) in data {
+            let src = MockErrExecutor::new();
+            let mut executor =
+                BatchLimitExecutor::new(src, limit, ExecSummaryCollectorDisabled).unwrap();
+            let res = executor.next_batch(get_rows);
+            if limit <= get_rows {
```

Comment:
how about no error for first several `next_batch()` calls and then there is an error?

Reply:
We already have this test case: ` (3, 4), // error happens after limit rows`